c/tskit/trees.c

```diff
@@ -4685,3 +4729,278 @@ tsk_tree_kc_distance(tsk_tree_t *self, tsk_tree_t *other, double lambda, double
+}
+
+static int
+update_kc_incremental(tsk_tree_t *self, kc_vectors *kc, tsk_edge_list_t *edges_out,
+    tsk_edge_list_t *edges_in)
+{
+    int ret = 0;
+    tsk_edge_list_t *e_list;
+    tsk_edge_t *e;
+    tsk_id_t u;
+    double root_time, time;
+    double *times = self->tree_sequence->tables->nodes.time;
+
+    for (e_list = edges_out; e_list != NULL; e_list = e_list->next) {
+        e = &e_list->edge;
+        u = e->child;
+        root_time = times[tsk_tree_node_root(self, u)];
+        ret = update_in_subtree_pairs(self, kc, u, root_time);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+    for (e_list = edges_in; e_list != NULL; e_list = e_list->next) {
+        e = &e_list->edge;
+        u = e->child;
+
+        if (tsk_tree_is_sample(self, u)) {
+            time = tsk_tree_branch_length(self, u);
+            update_kc_vectors_single_leaf(self->tree_sequence, kc, u, time);
+        }
+
+        root_time = times[tsk_tree_node_root(self, u)];
+        update_in_out_subtree_pairs(self, kc, e, root_time);
+        ret = update_in_subtree_pairs(self, kc, u, root_time);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+out:
+    return ret;
+}
+
+int
+tsk_treeseq_kc_distance(
+    tsk_treeseq_t *self, tsk_treeseq_t *other, double lambda_, double *result)
+{
+    int i;
+    tsk_id_t n;
+    double left, span, total;
+    tsk_tree_t t1, t2;
+    kc_vectors kc1, kc2;
+    tsk_diff_iter_t diff_iter1, diff_iter2;
+    tsk_treeseq_t *treeseqs[2] = { self, other };
+    tsk_tree_t *trees[2] = { &t1, &t2 };
+    kc_vectors *kcs[2] = { &kc1, &kc2 };
+    tsk_diff_iter_t *diff_iters[2] = { &diff_iter1, &diff_iter2 };
+    tsk_edge_list_t *t1_edges_out, *t1_edges_in, *t2_edges_out, *t2_edges_in;
+    double t1_left, t1_right, t2_left, t2_right;
+    int ret = 0;
+
+    for (i = 0; i < 2; i++) {
+        memset(trees[i], 0, sizeof(tsk_tree_t));
+        memset(kcs[i], 0, sizeof(kc_vectors));
+        memset(diff_iters[i], 0, sizeof(tsk_diff_iter_t));
+    }
+
+    ret = check_kc_distance_tree_sequence_inputs(self, other);
+    if (ret != 0) {
+        goto out;
+    }
+
+    n = (tsk_id_t) self->num_samples;
+    for (i = 0; i < 2; i++) {
+        ret = tsk_tree_init(trees[i], treeseqs[i], TSK_SAMPLE_LISTS);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = tsk_diff_iter_init(diff_iters[i], treeseqs[i]);
+        if (ret != 0) {
+            goto out;
+        }
+        ret = kc_vectors_alloc(kcs[i], n);
+        if (ret != 0) {
+            goto out;
+        }
+    }
+
+    total = 0;
+    left = 0;
+
+    ret = tsk_tree_first(&t1);
+    assert(ret == 1);
+    ret = tsk_diff_iter_next(
+        &diff_iter1, &t1_left, &t1_right, &t1_edges_out, &t1_edges_in);
+    assert(ret == 1);
+    ret = update_kc_incremental(&t1, &kc1, t1_edges_out, t1_edges_in);
+    if (ret != 0) {
+        goto out;
+    }
+    while (tsk_tree_next(&t2)) {
```

Comment:
OK, looking at the rest of this, maybe is it simpler to have ``t1`` and ``t2`` than ``&trees[0]`` and ``&trees[1]``, so it's up to you whether we make this change. But we have to fix the memory overrun problem above.